Adding Support for Inovelli VZM31-SN Blue 2-1 Dimmer Switch

InovelliUSA opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=1497) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=1497) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=1497) it.</sub>

cbaumler commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,445 @@
+-- Copyright 2024 Inovelli
```
I believe this copyright will need to be changed to SmartThings.

InovelliUSA replied:
Thanks, I updated it.

greens commented on drivers/SmartThings/zigbee-switch/profiles/inovelli-vzm31-sn.yml:
```diff
@@ -0,0 +1,279 @@
+  - name: "parameter1"
+    title: "1. Dimming Speed (Remote)"
+    description: "This changes the speed that the light dims up when controlled from the hub. A setting of '0' turns the light immediately on. Increasing the value slows down the transition speed. Value is multiplied by 100ms. 
+  Default=25 (2500ms or 2.5s)"
+    required: false
+    preferenceType: number
+    definition:
+      minimum: 0
+      maximum: 126
+      default: 25
+  - name: "parameter2"
+    title: "2. Dimming Speed (Local)"
+    description: "This changes the speed that the light dims up when controlled at the switch. A setting of '0' turns the light immediately on. Increasing the value slows down the transition speed. Value is multiplied by 100ms. 
+  (i.e 25 = 2500ms or 2.5s) Default=127 (Sync with parameter 1)"
+    required: false
+    preferenceType: number
+    definition:
+      minimum: 0
+      maximum: 127
+      default: 127
+  - name: "parameter3"
+    title: "3. Ramp Rate (Remote)"
+    description: "This changes the speed that the light turns on when controlled from the hub. A setting of '0' turns the light immediately on. Increasing the value slows down the transition speed. Value is multiplied by 100ms. 
+  (i.e 25 = 2500ms or 2.5s) Default=127 (Sync with parameter 1)"
+    required: false
+    preferenceType: number
+    definition:
+      minimum: 0
+      maximum: 127
+      default: 127
+  - name: "parameter4"
+    title: "4. Ramp Rate (Local)"
+    description: "This changes the speed that the light turns on when controlled at the switch. A setting of '0' turns the light immediately on. Increasing the value slows down the transition speed. Value is multiplied by 100ms. 
+  (i.e 25 = 2500ms or 2.5s) Default=127 (Sync with parameter 3)"
+    required: false
+    preferenceType: number
+    definition:
+      minimum: 0
+      maximum: 127
+      default: 127
```
One suggestion I would make would just be to unify all 4 of these parameters into 1 that controls the transition speed in all scenarios.

InovelliUSA replied:
We actually have 8 settings for this functionality so I have already consolidated it somewhat. I was told to try to use the same settings as what is in our red series driver that is already in the repo:

https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/2125c846a09b5b041f636d82e0855b10689a5859/drivers/SmartThings/zwave-switch/profiles/inovelli-dimmer-power-energy.yml#L48
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/2125c846a09b5b041f636d82e0855b10689a5859/drivers/SmartThings/zwave-switch/profiles/inovelli-dimmer-power-energy.yml#L98
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/2125c846a09b5b041f636d82e0855b10689a5859/drivers/SmartThings/zwave-switch/profiles/inovelli-dimmer-power-energy.yml#L108
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/2125c846a09b5b041f636d82e0855b10689a5859/drivers/SmartThings/zwave-switch/profiles/inovelli-dimmer-power-energy.yml#L118
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/2125c846a09b5b041f636d82e0855b10689a5859/drivers/SmartThings/zwave-switch/profiles/inovelli-dimmer-power-energy.yml#L155

greens commented on drivers/SmartThings/zigbee-switch/profiles/inovelli-vzm31-sn.yml:
```diff
@@ -0,0 +1,279 @@
+        "0": "No (default)"
+        "1": "Yes"
+      default: 0
+  - name: "parameter12"
+    title: "12. Auto Off Timer"
+    description: "Automatically turns the switch off after this many seconds. When the switch is turned on a timer is started. When the timer expires the switch turns off.
+  0=Auto Off Disabled."
+    required: false
+    preferenceType: number
+    definition:
+      minimum: 0
+      maximum: 32767
+      default: 0
```
We'd prefer if users used our rules and automations for this sort of behavior.

InovelliUSA replied:
Sure no problem. I can remove it.

greens commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,414 @@
+  end
+end
+
+local function configuration_handler(driver, device, zb_rx)
+  for i,v in ipairs(zb_rx.body.zcl_body.attr_records) do
+    if (v.attr_id.value == 0x000D) then
+    elseif (v.attr_id.value == 0x0015) then
+    else
+    end
+  end
+end
```
this function doesn't seem to do anything

greens commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,414 @@
+  end
+end
+
+local version_handler = function(driver, device, value, zb_rx)
+end
```
nor this one. Is this intentional?

InovelliUSA replied:
Just left over after stripping functionality from the driver. It was left on accident.

greens commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,414 @@
+      end
+    end
+    device:send(cluster_base.read_attribute(device, data_types.ClusterId(0x0000), 0x4000))
+  else
+    -- "info_changed event duplicate detected. Not performing any actions."
```
I'd just remove this branch if you're not logging

greens commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,414 @@
+    local send_configuration = function()
+      dev:send(cluster_base.build_manufacturer_specific_command(
+            dev,
+            0xfc31,
+            0x01,
+            0x122f,
```
can you pull the cluster_id and mfg_code out and put them at the top of the file?

greens commented on drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:
```diff
@@ -0,0 +1,414 @@
+      [0x0000] = {
+        [0x4000] = version_handler
+      },
+	  [clusters.SimpleMetering.ID] = {
+        [clusters.SimpleMetering.attributes.InstantaneousDemand.ID] = power_meter_handler,
+        [clusters.SimpleMetering.attributes.CurrentSummationDelivered.ID] = energy_meter_handler
+      },
+      [clusters.ElectricalMeasurement.ID] = {
+        [clusters.ElectricalMeasurement.attributes.ActivePower.ID] = power_meter_handler
+      }
+    },
+    global = {
+      [0xFC31] = {
+        [0x01] = configuration_handler
+      }
+    },
+    cluster = {
+      [0xFC31] = {
+        [0x00] = scene_handler,
+      }
+    }
+  },
+  capability_handlers = {
+    [capabilities.switch.ID] = {
+      [capabilities.switch.commands.on.NAME] = on_handler,
+      [capabilities.switch.commands.off.NAME] = off_handler,
```
nit: some mixed tabs and spaces here

greens commented on drivers/SmartThings/zigbee-switch/profiles/child-notification.yml:
Could you rename this to match the other related profiles `rgbw-bulb-[min_color_temp]-[max_color_temp].yml` and add in the `firmwareUpdate` capability?

lelandblue commented:
Hey @InovelliUSA Can you check out the open comment above on this PR please at your nearest opportunity? Thank you!

InovelliUSA commented:
> Hey @InovelliUSA Can you check out the open comment above on this PR please at your nearest opportunity? Thank you!

Sure, I will check it out right now.

InovelliUSA commented:
I believe I have adjusted everything requested. I did find a bug that I addressed in info_changed as well. Other than that, I just changed what was asked.

https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1497/commits/8d4ab6a452439431d36250e2626b97ba2725643d

greens commented on drivers/SmartThings/zigbee-switch/profiles/child-notification.yml:
you need to `git rm` this file

lelandblue commented:
Hey @InovelliUSA Thanks so much. 

We noticed that there arent yet Unit Tests written for this driver. We should have called this out a bit sooner, so apologies. I will recommend that we continue with the WWST process and we can review the driver tests in parallel.  To learn more about unit driver tests, please see https://developer.smartthings.com/docs/devices/hub-connected/test-your-driver

greens commented:
> ```
> Checking drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua 1 warning
> 
>     drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:62:9: (W211) unused variable numeric
> ```

This still needs to be fixed in the meantime as well.

InovelliUSA commented:
> > ```
> > Checking drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua 1 warning
> > 
> >     drivers/SmartThings/zigbee-switch/src/inovelli-vzm31-sn/init.lua:62:9: (W211) unused variable numeric
> > ```
> 
> This still needs to be fixed in the meantime as well.

This should be fixed now.

lelandblue commented:
> Hey @InovelliUSA Thanks so much.
> 
> We noticed that there arent yet Unit Tests written for this driver. We should have called this out a bit sooner, so apologies. I will recommend that we continue with the WWST process and we can review the driver tests in parallel. To learn more about unit driver tests, please see https://developer.smartthings.com/docs/devices/hub-connected/test-your-driver

Hey @InovelliUSA - Can you let me know if you have any questions on the above message please? Thank you.

InovelliUSA commented:
> > Hey @InovelliUSA Thanks so much.
> > We noticed that there arent yet Unit Tests written for this driver. We should have called this out a bit sooner, so apologies. I will recommend that we continue with the WWST process and we can review the driver tests in parallel. To learn more about unit driver tests, please see https://developer.smartthings.com/docs/devices/hub-connected/test-your-driver
> 
> Hey @InovelliUSA - Can you let me know if you have any questions on the above message please? Thank you.

I have always tested with the device itself, so I haven't created the unit tests. I will have to research how to do that.

lelandblue commented:
@InovelliUSA Can you please review and sign the  CLA? Thank you!

lelandblue commented:
Hello - I am going to close this PR. If it needs to be reopened, it can be. Thank you.

lelandblue commented:
Hello @InovelliUSA We can merge this PR and deploy, however we need a signature on the CLA. Can you please review at your nearest opportunity? 

 Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=1497) before we can accept your contribution.

InovelliUSA commented:
> Hello @InovelliUSA We can merge this PR and deploy, however we need a signature on the CLA. Can you please review at your nearest opportunity?
> 
> Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SmartThingsCommunity/SmartThingsEdgeDrivers?pullRequest=1497) before we can accept your contribution.

Sorry about that, I think I just completed it.

lelandblue commented:
It appears that @erocm123 needs to still sign, are they available to sign the CLA?
